arch/arm/core/aarch32/prep_c.c

```diff
@@ -40,43 +40,53 @@
 #ifdef CONFIG_CPU_CORTEX_M_HAS_VTOR
 
 #ifdef CONFIG_XIP
-#define VECTOR_ADDRESS ((uintptr_t)_vector_start)
+#ifdef CONFIG_SW_VECTOR_RELAY
+#define VECTOR_ADDRESS ((uintptr_t)__vector_relay_table)
 #else
-#define VECTOR_ADDRESS CONFIG_SRAM_BASE_ADDRESS
+#define VECTOR_ADDRESS ((uintptr_t)_vector_start)
 #endif
-static inline void relocate_vector_table(void)
-{
-	SCB->VTOR = VECTOR_ADDRESS & SCB_VTOR_TBLOFF_Msk;
-	__DSB();
-	__ISB();
-}
+#else /* CONFIG_XIP */
+#define VECTOR_ADDRESS CONFIG_SRAM_BASE_ADDRESS
+#endif /* CONFIG_XIP */
 
-#else
+#else /* CONFIG_CPU_CORTEX_M_HAS_VTOR */
+
+#define VECTOR_ADDRESS 0
+
+#endif /* CONFIG_CPU_CORTEX_M_HAS_VTOR */
 
 #if defined(CONFIG_SW_VECTOR_RELAY) || defined(CONFIG_SW_VECTOR_RELAY_CLIENT)
 Z_GENERIC_SECTION(.vt_pointer_section) __attribute__((used))
 void *_vector_table_pointer;
 #endif
 
-#define VECTOR_ADDRESS 0
 
+#ifdef CONFIG_CPU_CORTEX_M_HAS_VTOR
+static inline void relocate_vector_table(void)
+#else
 void __weak relocate_vector_table(void)
+#endif
 {
+#if defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) && !defined(CONFIG_SW_VECTOR_RELAY_CLIENT)
+	SCB->VTOR = VECTOR_ADDRESS & SCB_VTOR_TBLOFF_Msk;
```

Comment:
I realize this was just moved from the code above, but I don't see how this would work. If the VECTOR_ADDRESS is not aligned according to the mask, it just won't work. Masking off the low bits won't help that (in fact, the hardware will happily do that for you when you write the value).

Reply:
I agree, at least on Cortex-M4, the hardware does it for you.
From Cortex-M4 Devices Generic User Guide
>Table alignment requirements mean that bits[6:0] of the table offset are always zero.

Reply:
It isn't as much a matter of whether the mask is there or not, but that if the masked out bits are set, there isn't any hope of the vector table working. The HW doesn't add the VTOR value to the given address, it just uses the upper bits from the mask, and uses the lower bits from the vector. In other words, it doesn't matter whether you mask it or not.

The alignment depends on the particular SoC, based on the number of IRQ vectors used in that device.

It might be nice, at compile time, to detect the table being not at an aligned address.

Reply:
>The alignment depends on the particular SoC, based on the number of IRQ vectors used in that device.

>It might be nice, at compile time, to detect the table being not at an aligned address.

@d3zd3z this is addressed now in #26800 . In fact this was missing in the code. Now we ensure at compile-time the vector table is properly aligned.